File: packages/core/src/storage/InMemoryProvider.js
import { LRUCache } from 'lru-cache';
import StorageProvider from './StorageProvider.js';

export default class InMemoryProvider extends StorageProvider {
  constructor(options) {
    super(options);
    
    const cacheOptions = {
      max: options.maxSize || 50 * 1024 * 1024, // 50MB default
      ttl: options.ttl ? options.ttl * 1000 : undefined, // TTL in milliseconds
      // Calculate size of buffer for max size enforcement
      sizeCalculation: (value) => value.length,
    };

    this.cache = new LRUCache(cacheOptions);
  }

  async save(fileName, data) {
    this.cache.set(fileName, data);
    // For in-memory, the "path" is just the key
    return `in-memory://${fileName}`;
  }

  // The LRU cache handles its own cleanup, so this method is a no-op
  // but is here to fulfill the interface contract.
  async cleanup() {
    console.log('InMemoryProvider uses an LRU cache for automatic cleanup.');
    return Promise.resolve();
  }

  get(fileName) {
    return this.cache.get(fileName);
  }
}

File: python/pagesnap/storage/in_memory_provider.py
from .storage_provider import StorageProvider
from lru import LRU

class InMemoryProvider(StorageProvider):
    def __init__(self, options: dict):
        super().__init__(options)
        # lru-dict uses max_size in number of items, not bytes.
        # This is a slight deviation from the JS implementation but achieves the same goal.
        self.cache = LRU(options.get("max_items", 256))

    async def save(self, file_name: str, data: bytes) -> str:
        self.cache[file_name] = data
        return f"in-memory://{file_name}"

    def get(self, file_name: str) -> bytes:
        return self.cache.get(file_name)
